### PR TITLE
JENKINS-37104# Matrix project handling

### PR DESCRIPTION
**Pull request by vivek**
# Description

See https://issues.jenkins-ci.org/browse/JENKINS-37104

# Submitter checklist
- [x] Link to JIRA ticket in description, if appropriate.
- [x] Change is code complete and matches issue description
- [x] Apppropriate unit or acceptance tests or explaination to why this change has no tests
- [x] Reviewer's manual test instructions provided in PR description. See Reviewer's first task below.

- Create a matrix project mp1 with two axis with two values each and run it, you should get total of 4 builds, one for each combination and there should be build for mp1 as well.
- Do curl -v http://localhost:8080/jenkins/blue/rest/search/?q=type:pipeline;excludedFromFlattening:jenkins.branch.MultiBranchProject,hudson.matrix.MatrixProject to get list of pipelines, here we are asking not to inline MatrixProject children.
- The list of pipeline should have mp1 project with self href pointing to "/job/mp1/"
- curl -v http://localhost:8080/jenkins/blue/rest/classes/io.jenkins.blueocean.rest.impl.pipeline.MatrixProjectImpl/ should give list of capabilities that includes hudson.matrix.MatrixProject

- [x] Ran Acceptance Test Harness against PR changes.
https://ci.blueocean.io/job/Acceptance%20Tests%20Param/19/, it failed with unrelated error to download jquery2:

    [ERROR - 2016-08-17T01:17:57.972Z] Session [0f406550-6418-11e6-9fd4-d5cee85dbc1a] - page.onError - msg: Module load failure: Timed out waiting on module 'jquery-detached:jquery2' to load.

Triggered another: https://ci.blueocean.io/job/Acceptance%20Tests%20Param/20/

# Reviewer checklist
- [ ] Run the changes and verified the change matches the issue description
- [x] Reviewed the code
- [x] Verified that the appropriate tests have been written or valid explaination given

@jenkinsci/code-reviewers @reviewbybees 

- Bare minimum support for Matrix project. Ensures it doesn't break.
- For matrix project, self href points to classic URL path ("/job/:name")
- Adds hudson.matrix.MatrixProject capability
- UI to check hudson.matrix.MatrixProject capability and if present create a link using returned self href,
  so that when user clicks on it, he gets redirected to classic Matrix job page.

### CODE REVIEW

**Comment by michaelneale** [comment]
@vivek seems to have upset the favoriting feature (cc @cliffmeyers) - kicking it again to see if sporadic.

**Comment by michaelneale** [comment]
🐛  fails the fav. tests with 500 errors, but not sure why. Can you get @cliffmeyers to cast eye over it?

**Comment by vivek** [comment]
@michaelneale I removed test dependency on matrix project plugin thinking we don't need it. This broke the tests as Favorite plugin depends on matrix project plugin. I reverted that dependency and tests pass now. Is there anything holding back this merge?

**Comment by michaelneale** [comment]
@vivek I think it should be ok - myself or someone else just needs to eyeball it (but I won't get to it this morning). If anyone else can try this one out and check it is ok, would be appreciated.

**Comment by imeredith** [comment]
@vivek I cant test out the code right now, but where is the pick that makes the following work?

* UI to check hudson.matrix.MatrixProject capability and if present create a link using returned self href, so that when user clicks on it, he gets redirected to classic Matrix job page.

**Comment by vivek** [comment]
@imeredith 'but where is the pick that makes the following work?' you mean mockup?

**Comment by imeredith** [comment]
@vivek Sorry i meant 'part'. I dont see code that makes UI redirect if matrix job is clicked?

**Comment by vivek** [comment]
@imeredith Oh, its to be done in separate ticket, not part of this PR, https://issues.jenkins-ci.org/browse/JENKINS-37427.

**Comment by michaelneale** [comment]
@vivek ready for review (can we remove the work-in-progress label?)

**Comment by imeredith** [comment]
Ok 🐝  from me then.

**Comment by vivek** [comment]
@imeredith I take that as ready to merge?

**Comment by michaelneale** [comment]
I won't have a chance to eyeball this today, so if @imeredith can or @vivek you are happy it loads up blueocean ui with some existing jobs etc - then I am ok with it. Code looks good. ATH passes.

**Comment by vivek** [comment]
@michaelneale yes I have done a clean build, loads up fine with different kind of jobs. Going to merge.

File: blueocean-pipeline-api-impl/src/main/java/io/jenkins/blueocean/rest/impl/pipeline/MatrixProjectImpl.java [hunk 1]
```diff
@@ -0,0 +1,101 @@
+package io.jenkins.blueocean.rest.impl.pipeline;
+
+/**
+ * @author Vivek Pandey
+ */
+
+import hudson.Extension;
+import hudson.matrix.MatrixProject;
```
Comment (basil) [inline blueocean-pipeline-api-impl/src/main/java/io/jenkins/blueocean/rest/impl/pipeline/MatrixProjectImpl.java:8]: Neglected to add the corresponding dependency to `blueocean-pipeline-api-impl/pom.xml`, implicitly relying on `htmlpublisher` to provide it (no longer the case after https://github.com/jenkinsci/htmlpublisher-plugin/pull/287).

**Pull request merged**